llvm/include/llvm/MCA/HardwareUnits/ResourceManager.h

```diff
@@ -430,7 +430,7 @@ class ResourceManager {
 
   void issueInstruction(
       const InstrDesc &Desc,
-      SmallVectorImpl<std::pair<ResourceRef, ReleaseAtCycles>> &Pipes);
+      SmallVectorImpl<std::pair<ResourceRef, NumCyclesUsed>> &Pipes);
```

Comment:
How comes that LLVM-MCA does not need to know about the AcquireAtCycle?

Reply:
NumCycleUsed is based on the `ReleaseAtCycle - AcquireAtCycle`. In that sense, it does know about AcquireAtCycle.

I think what you are pointing out here is that the current implementation will use the resource for the correct number of cycles now, but it ignores when it can start acquiring the resource relative to when the instruction is issued. To demonstrate, say an instruction X uses resource A with (Acquire, Release) = (0, 2) and resource B with (Acquire, Release) = (1, 2). This patch will model that resource A is used for 2 cycles and resource B for 1 cycle, but it does not model that we have to wait one cycle before consuming resource B.

I should look into adding this support. Thanks for catching this!

Reply:
As Michael wrote, this patch doesn't add support for `AcquireAtCycle`.

The difficult part will be refactoring the ResourceManager logic so to keep track of multiple `AcquireAtCycle` requests, and quickly resolve checks on whether resources are busy or not.
Changing that logic might not be too complicated. However, doing it without slowing down the process too much might be challenging.

At the moment, mca doesn't know about the `AcquireAtCycles`, so this will be a problem for future patches.